Add MAEST (Music Audio Efficient Spectrogram Transformer) model

palonso opened the pull request:
# What does this PR do?
This PR implements the MAEST models (based on AST) as proposed in #26491.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR. 

Given that MAEST is intended for music applications and is audio-based, I consider that @sanchit-gandhi could be the appropriate member for having a look at it.

Thank you very much for considering this PR!

palonso committed:
Add MAEST implementation

The MAEST model is based on AST (audio_spectrogram_transformer) but it features a custom
mel-spectrogram input representation optimized for musical purposes.
While MAEST was trained using Essentia mel-spectrograms, this library does not support
multi-threading, making it unsuitable for the implementation of the FeatureExtractor. For now, the
implementation relies in Librosa, since similar mel-spectrogram configuration can be achieved,
producing close to identical results.

palonso committed:
Add MAEST changes related to the automatic configuration

palonso committed:
Add MAEST-related doc

palonso committed:
Add MAEST-related tests

palonso committed:
Remove SWA version of the models

We decided to provide the SWA version of the models by default since they performed better in all
the evaluations. We'll consider uploading the non-swa version of the models too if someone requires
them.

palonso committed:
Add MAEST to dummy_pt_objects.py

palonso commented:
Setting this PR as a draft, for now, since several tests are not passing locally on `test_modeling_maest.py`. Similarly, I found that the same tests are not passing for `test_modeling_audio_spectrogram_transformer`.py, so I'll need to investigate it.

palonso committed:
Add MAEST to dummy_speech_objects.py

palonso committed:
Fix import order

palonso committed:
Remove MAEST classes overlapping AST

MAESTModel, and MAESTForAudioClassification should be removed
according to #26491 since they follow the same implementation as
AST.

palonso committed:
Fix MAEST tests

- Remove PipelineTesterMixin since MAEST is not attached to a
any pipeline right now.

- Add missing `setUp` and `test_config` methods.

palonso committed:
Discard MAESTConfig class

Since this class is neither required and the same results can be
achieved using ASTConfig.

palonso committed:
Remove missing MAESTConfig import

palonso committed:
Fix MAESTFeatureExtractor style

palonso committed:
Restore var MAEST_PRETRAINED_CONFIG_ARCHIVE_MAP

sanchit-gandhi commented on src/transformers/models/maest/feature_extraction_maest.py:
```diff
@@ -0,0 +1,148 @@
+logger = logging.get_logger(__name__)
+
+
+class MAESTFeatureExtractor(ASTFeatureExtractor):
```
Rather than subclassing `ASTFeatureExtractor`, it's probably cleaner if you just use `# Copied from` for each class that is a direct copy from AST, and override the classes that are new

palonso replied:
Addressed in 9ce296f

sanchit-gandhi commented on src/transformers/models/maest/feature_extraction_maest.py:
```diff
@@ -0,0 +1,148 @@
+
+import numpy as np
+import torch
+from librosa.feature import melspectrogram
```
We have an `audio_utils` file that allows you to run Mel spectrograms without a `librosa` requirement. Here's an example for the SpeechT5 model: https://github.com/huggingface/transformers/blob/9ed538f2e67ee10323d96c97284cf83d44f0c507/src/transformers/models/speecht5/feature_extraction_speecht5.py#L161

In `transformers`, we try to avoid extra dependencies (like `librosa`) unless seriously required. Here, I think we can bypass the requirement by using `audio_utils`

palonso replied:
Addressed in 57324bf

sanchit-gandhi commented on src/transformers/models/maest/modeling_maest.py:
```diff
@@ -0,0 +1,42 @@
+# coding=utf-8
```
As mentioned by @NielsRogge - we can skip adding a modelling file since it's one-to-one the same as AST, and promote the `AutoModel` API

palonso replied:
Addressed in ea1c43e

palonso committed:
Improve MAEST doc

palonso committed:
Use audio_utils in MAESTFeatureExtraction

This discards the dependency on librosa as proposed in
https://github.com/huggingface/transformers/pull/26496#discussion_r1342948616

palonso committed:
Do not inherit from ASTFeatureExtractor

as proposed in https://github.com/huggingface/transformers/pull/26496#discussion_r1342947027

palonso committed:
Add missing functionalities to MAESTFeatureExtractor

- Similar to ASTFeatureExtractor, it should inherit from SequenceFeatureExtractor
- Implement `normaize` and `__call__` methods (copied from ASTFeatureExtractor)

palonso committed:
Fix typo in variable name

palonso committed:
Store MAESTFeatureExtractor parameters as lists

The MAESTFEatureExtractor parameters `mel_fb` (mel-filter-bank) and window
(window functions) are computed only once on the class initialization for
efficiency. Storing them as numpy arrays causes `test_feat_extract_to_json_string`
and `test_feat_extract_to_json_file` to fail because self.assertEqual is not
intended to operate on numpy matrices matrices (you need to call `all()` or `any()`
on top).

Casting these variables to Python lists fixes the issue.

palonso committed:
Update MAESTFeatureExtractor doc

palonso committed:
Use dedicated hop_length variable

palonso committed:
Fix incorrect transposition in MAESTFeatureExtractor

palonso committed:
Remove modeling_maest.py and all references

As mentioned here https://github.com/huggingface/transformers/pull/26496#discussion_r1342949343

palonso committed:
Merge main branch

palonso committed:
Fix MAEST documentation

palonso committed:
Fill missing doc in FeatureExtractorMAEST
